c10/util/Float8_e4m3.h

```diff
@@ -0,0 +1,247 @@
+#pragma once
+
+/// Defines the Float8_e4m3 type (8-bit floating-point) including conversions
+/// to standard C types and basic arithmetic operations. Note that arithmetic
+/// operations are implemented by converting to floating point and
+/// performing the operation in float32.
+/// Binary configuration:
+/// s eeee mmm
+/// 1 sign bit
+/// 4 exponent bits
+/// 3 mantissa bits
+/// bias = 7
+///
+/// Implementation based on the paper https://arxiv.org/pdf/2209.05433.pdf
+/// and inspired by Half implementation from pytorch/c10/util/Half.h
+
+#include <c10/macros/Macros.h>
+#include <c10/util/C++17.h>
+#include <c10/util/TypeSafeSignMath.h>
+#include <c10/util/floating_point_utils.h>
+#include <type_traits>
+
+#if defined(__cplusplus) && (__cplusplus >= 201103L)
+#include <cmath>
+#include <cstdint>
+#elif !defined(__OPENCL_VERSION__)
+#include <math.h>
+#include <stdint.h>
+#endif
+
+#ifdef _MSC_VER
+#include <intrin.h>
+#endif
+
+#include <cstdint>
+#include <cstring>
+#include <iosfwd>
+#include <limits>
+#include <sstream>
+#include <stdexcept>
+#include <string>
+#include <utility>
+
+// Standard check for compiling CUDA with clang
+#if defined(__clang__) && defined(__CUDA__) && defined(__CUDA_ARCH__)
+#define C10_DEVICE_HOST_FUNCTION __device__ __host__
+#else
+#define C10_DEVICE_HOST_FUNCTION
+#endif
+
+#include <typeinfo> // operator typeid
+
+namespace c10 {
+
+namespace detail {
+
+/*
+ * Convert a 8-bit floating-point number in fp8 E4M3 format, in bit
+ * representation, to a 32-bit floating-point number in IEEE single-precision
+ * format, in bit representation.
+ *
+ * @note The implementation doesn't use any floating-point operations.
+ */
+inline float fp8e4m3_to_fp32_value(uint8_t input) {
```

Comment:
nice! I didn't read line by line yet, but high level question - what's your level of confidence on the correctness of these functions?  Have these been matched up with NVIDIA's casts or https://github.com/pytorch/FBGEMM/blob/277677039bae25b2570a73013b03bfaa9d2a523e/fbgemm_gpu/include/fbgemm_gpu/quantize_ops_utils.h ?

Reply:
It was quite extensively tested against Gaudi2, which supports these variants natively. However, in next commits, I'll add comparative tests for all bits configurations against some reference implementation.
Can you guarantee that fbgemm implementation is fully aligned with the specification? Also, do you know if there is some open fp8_e5m2 reference implementation I could use for testing?

Reply:
the fbgemm implementation has been used extensively at Meta but it hasn't been matched up against real hardware afaik.

Your test cases look good, I'd say if future changes on numerics are needed that can be done in future PRs and not block this one.  LGTM.